### ISSUE TITLE
Native package with native modules

### ISSUE THREAD

**Issue by 0x80** [jxcore/jxcore#370]
I've been trying to make a native package bundle which includes a native module (serialport). I find only information on how to make this work by excluding the node_modules folder in the package and use a preinstall step, but this would require the user to have jxcore installed if I understand correctly.

I need a selfcontained executable I can distribute, so currently I am out of luck right? Just trying to clear this up.

**Comment by ktrzeciaknubisa** [comment]
Well, you might be right. If this wouldn't be a `-native` package, you could set `extract : true` in jxp file (or call `jx package` with `-extract` switch). But apparently native packages do not support `extract` option. @obastemur I cannot recall right now - is this by design or we have just missed that?

**Comment by obastemur** [comment]
@ktrzeciaknubisa this is intentional. I did mark this issue with an enhancement label. We can make native module embedding part jx package aware. So it could extract the node file from the package on request. 

@0x80 You can workaround this problem; 
 - Check if the .node file is extracted to the path 
 - use fs.read etc to read .node file and write it back to disk

**Comment by ktrzeciaknubisa** [comment]
Yes, but after second thought that would work only on same platform as where we created the package... If the app is supposed to be truly cross platform, native modules should be reinstalled on target machines, right?

@0x80 how about using preinstall steps also for downloading jx binary to app directory?

**Comment by obastemur** [comment]
I don't think he is planning to create a native package for Windows 64 and run it on OSX or Ubuntu though.

**Comment by ktrzeciaknubisa** [comment]
> We can make native module embedding part jx package aware. So it could extract the node file from the package on request. 

This sounds nice. I also thought to extend `extract` option from boolean value to string (file names, comma separated like in case of `-add` or `-slim`, also with wildcards) so only the specified files would be extracted. 

If `extract == "assets"` then js source files will not be extracted at all. 

How about that?

**Comment by obastemur** [comment]
IMHO JS Object is the best option. A premature design suggestion;
```
extract: {
  from: [ folders etc. here],
  override: true/false,
  custom_message: "show this message when extracting"
  verbose: true/false -> show/hide list of files while extracting,
  custom_pre_actions: [ cmds to run before extration ],
  custom_after_actions: [cmds to run after extraction]  
}
```

**Comment by ktrzeciaknubisa** [comment]
Good idea. Though I'm concerned with `override`. The existing folder may contain completely something else than user's app (just having the same name). Maybe we should mark the extracted folder e.g. by writing into it `test.extracted` so each consecutive extract will know, that the folder may be safely overridden.

**Comment by 0x80** [comment]
Hi guys! Thanks for the response. I was indeed planning to package the app seperately for each platform. However in the meantime I've been playing with the Electron shell and decided to go with that since it better fits my current project. I was planning the app without a user interface but we decided we need one after all...

**Comment by ktrzeciaknubisa** [comment]
@obastemur Let me share with some thoughts on the subject:

1. Partial files extracting would make sense only if we would extract directly to folder, where the package is located (without creating separate directory for package contents). Otherwise unextracted js files would have trouble to find extracted assets on file system.

2. Right now after the successful extract, the extracted app is executed by a package with `cmdSync` (rather than executing the package itself). This will not work with native packages, as we have no jx executable available. My suggestion is to always execute startup file from a package, (no matter if it's extracted or not). This way native packages may become extractable (I already tested this).

But in order to have 2. working, we need to solve 1. (get rid of extra subfolder for extracted directory contents). Would that be ok?

**Comment by ozzyisti** [comment]
Hi! I'm also in this situation, trying to compile a native package bundle which includes native modules (serialport and socket.io). @ktrzeciaknubisa you mentioned the possibility to using preinstall steps also for downloading jx binary to app directory.
Could you provide me an example jxp file for this?
Is this really possible? Or there is another solution on which somebody is already working?

**Comment by ktrzeciaknubisa** [comment]
HI @ozzyisti What I had on mind was using e.g. curl to download a jx package and then unzipping it. For example I've created a package this way: 

```bash
jxv8 package test.js -add -native -preinstall "curl https://jxcore.s3.amazonaws.com/0301/jx_osx64v8.zip -o file.zip,unzip -o -j file.zip"
```

And after executing `./test`, the `preInstall` steps did that for me.

Inside `test.jxp` it may look this way:

```js
	"preInstall": [
		"curl https://jxcore.s3.amazonaws.com/0301/jx_osx64v8.zip -o file.zip",
		"unzip -o -j file.zip"
	],
```

But I admit, this is not very convenient at the moment (you need to know an exact url and exact platform (e.g. osx64v8). I'm thinking to add an option for `curl http://jxcore.com/xil.sh | bash` installation script to install the binary locally instead of in global path.

**Comment by ozzyisti** [comment]
Thanks for the answer @ktrzeciaknubisa. Unfortunately this solution isn't going to work on Windows machines as I cannot be sure if there is curl installed or any other command line solution for downloading jx.zip. On Linux I would use 'wget', but on Windows there is no easy solution for this step, AFAIK.

**Comment by ktrzeciaknubisa** [comment]
@ozzyisti We may provide a simple GUI/command web installer for windows soon, so it should solve your problems. 

As for unix platforms, the installation script is updated and supports local install, too (to the current directory):

```bash
$ curl http://jxcore.com/xil.sh | bash -s local
```
I'm also going to document this.

**Comment by ozzyisti** [comment]
Nice @ktrzeciaknubisa ! I can not wait to try it!

**Comment by ktrzeciaknubisa** [comment]
The first and most important part for partial packages extraction is implemented with 44befcb7a344aac651a69174ff25fe522f07452c. By using this feature we can mark e.g. *.node* files as extractable thus run packages without any additional steps.

```bash
$ jx package myapp.js --extract-what "*.node" --extract-app-dir
$ jx myapp.jx
```

The same also works for native packages, but not on Windows yet (due to #489).

```bash
$ jx package myapp.js --extract-what "*.node" --extract-app-dir -native
$ ./myapp
```

That commit doesn't exhaust all proposed extract features, but this is good start. First of all extraction for native packages was disabled at all.

Documentation updated [here](https://github.com/jxcore/jxcore/blob/0.3.0.6/doc/api/jxcore-feature-packaging-code-protection.markdown#-extract).
